sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,1013 @@
+                    # sample constraints if they are too many
+                    if len(imp_ind) > self.max_constraints:
+                        imp_ind = self.random_state_.choice(
+                            imp_ind, self.max_constraints, replace=False)
+
+                    dims = (len(ind_out), len(ind_in))
+                    ii, jj = np.unravel_index(imp_ind, dims=dims)
+                    imp_row = ind_out[ii]
+                    imp_col = ind_in[jj]
+                    new_imp = csr_matrix((np.ones(len(imp_row), dtype=np.int8),
+                                          (imp_row, imp_col)), dtype=np.int8,
+                                         shape=(n_samples, n_samples))
+                    impostors_sp = impostors_sp + new_imp
+
+            impostors_sp = impostors_sp.tocoo(copy=False)
+            imp_row = impostors_sp.row
+            imp_col = impostors_sp.col
+            imp_dist = paired_distances_batch(X_embedded, imp_row, imp_col)
+        else:
+            # Initialize impostors vectors
+            imp_row, imp_col, imp_dist = [], [], []
+            for class_id in self.classes_inverse_non_singleton_[:-1]:
+                ind_in, = np.where(y == class_id)
+                ind_out, = np.where(y > class_id)
+
+                # Subdivide ind_out x ind_in to chunks of a size that is
+                # fitting in memory
+                imp_ind, dist_batch = _find_impostors_batch(
+                    X_embedded[ind_out], X_embedded[ind_in],
+                    margin_radii[ind_out], margin_radii[ind_in],
+                    return_distance=True)
+
+                if len(imp_ind):
+                    # sample constraints if they are too many
+                    if len(imp_ind) > self.max_constraints:
+                        ind_sampled = self.random_state_.choice(
+                            len(imp_ind), self.max_constraints, replace=False)
+                        imp_ind = imp_ind[ind_sampled]
+                        dist_batch = dist_batch[ind_sampled]
+
+                    dims = (len(ind_out), len(ind_in))
+                    ii, jj = np.unravel_index(imp_ind, dims=dims)
+                    ind_out_batch = ind_out[ii]
+                    ind_in_batch = ind_in[jj]
+                    try:
+                        imp_row.extend(ind_out_batch)
+                        imp_col.extend(ind_in_batch)
+                        imp_dist.extend(dist_batch)
+                    except TypeError:
+                        imp_row.append(ind_out_batch)
+                        imp_col.append(ind_in_batch)
+                        imp_dist.append(dist_batch)
+
+            imp_row = np.asarray(imp_row)
+            imp_col = np.asarray(imp_col)
+            imp_dist = np.asarray(imp_dist)
+
+        return imp_row, imp_col, imp_dist
+
+
+##########################
+# Some helper functions #
+#########################
+
+
+def select_target_neighbors(X, y, n_neighbors, algorithm='auto', n_jobs=1,
```

Comment:
make it private?

Reply:
keeping it private for now and exposing it if there is a clear need is easier than the other way around

Reply:
Ok.